test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+      // forceId is set to false for the purpose of updating the same affected user within the
+      // `Email Update` test cases.
+      forceId: false,
+      // Speed up the password hashing algorithm for tests
+      saltWorkFactor: 4,
+    });
+    app.model(User2, {dataSource: 'db'});
+
+    AccessToken = app.registry.getModel('AccessToken');
+    app.model(AccessToken, {dataSource: 'db'});
+
+    // Update AccessToken and Users relations to use polymorphism
+    AccessToken.belongsTo('user', {idName: 'id', polymorphic: {idType: 'string',
+      foreignKey: 'userId', discriminator: 'principalType'}});
+    User.hasMany('accessTokens', {polymorphic: {foreignKey: 'userId',
+      discriminator: 'principalType'}});
+    User2.hasMany('accessTokens', {polymorphic: {foreignKey: 'userId',
+      discriminator: 'principalType'}});
+
+    // Speed up the password hashing algorithm
+    // for tests using the built-in User model
+    User.settings.saltWorkFactor = 4;
+
+    // allow many User.afterRemote's to be called
+    User.setMaxListeners(0);
+
+    app.enableAuth({dataSource: 'db'});
+    app.use(loopback.token({model: AccessToken}));
+    app.use(loopback.rest());
+
+    User.create(validCredentials, function(err, user) {
+      if (err) return done(err);
+      validCredentialsUser = user;
+      User2.create(validCredentials, function(err, user2) {
+        if (err) return done(err);
+        validCredentialsUser2 = user2;
+        done(null, user, user2);
+      });
+    });
+  });
+
+  describe('User.login', function() {
+    it('Login a user from 1st user model by providing credentials', function(done) {
+      User.login(validCredentials, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+
+        done();
+      });
+    });
+
+    it('Login a user from 2nd user model by providing credentials', function(done) {
+      User2.login(validCredentials, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser2);
+
+        done();
+      });
+    });
+
+    it('Login a user by providing email credentials (email case-sensitivity off)', function(done) {
+      User.settings.caseSensitiveEmail = false;
+      User.login(validMixedCaseEmailCredentials, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+
+        done();
+      });
+    });
+
+    it('Try to login with invalid email case', function(done) {
+      User.login(validMixedCaseEmailCredentials, function(err, accessToken) {
+        assert(err);
+
+        done();
+      });
+    });
+
+    it('Login a user by providing credentials with TTL', function(done) {
+      User.login(validCredentialsWithTTL, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+        assert.equal(accessToken.ttl, validCredentialsWithTTL.ttl);
+
+        done();
+      });
+    });
+
+    it('honors default `createAccessToken` implementation', function(done) {
```

Comment:
ditto, this is IMO not related to polymorphic users and can be removed. The fact that an access token can be created was already verified by `login` tests above.